Context.php

```diff
@@ -0,0 +1,326 @@
+ *
+ * New BSD License
+ *
+ * Copyright © 2007-2015, Hoa community. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of the Hoa nor the names of its contributors may be
+ *       used to endorse or promote products derived from this software without
+ *       specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS AND CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+namespace Hoa\Math;
+
+use Hoa\Math;
+
+/**
+ * Class \Hoa\Math\Context.
+ *
+ * @copyright  Copyright © 2007-2015 Hoa community
+ *             Ivan Enderlin, Cédric Dugat.
+ * @license    New BSD License
+ */
+class Context
+{
+    /**
+     * List of supported functions: identifier => values as callable
+     *
+     * @var \ArrayObject
+     */
+    protected $_functions = null;
+
+    /**
+     * List of supported constants
+     *
+     * @var \ArrayObject
+     */
+    protected $_constants = null;
+
+    /**
+     * List of supported variables
+     *
+     * @var \ArrayObject
+     */
+    protected $_variables = null;
+
+    /**
+     * Initialize constants and functions.
+     *
+     * @return  void
+     */
+    public function __construct()
+    {
+        $this->initializeConstants();
+        $this->initializeFunctions();
+        $this->initializeVariables();
+
+        return;
+    }
+
+    /**
+     * Add a variable.
+     *
+     * @param   string   $name     Variable name.
+     * @param   callable $callable Callable.
+     * @return  void
+     */
+    public function addVariable($name, callable $callable)
+    {
+        $this->_variables[$name] = xcallable($callable);
+
+        return;
+    }
+
+    /**
+     * Get variables.
+     *
+     * @return \ArrayObject
+     */
+    public function getVariables()
```

Comment:
Actually, we use to declare methods in this order:
* add/set,
* get,
* get all.
So `getVariables` should move after `getVariable` declaration, please :-). Same for other methods.